nilearn/region_extraction/regionextractor.py

```diff
@@ -0,0 +1,232 @@
+        the regions within which have voxels survived more than "min_size".
+
+    threshold: float or string, default string {'auto'}, optional
+        chooses to extract the foreground objects. If float, this value is
+        directly taken as a threshold or 'auto' option chooses automatically.
+        We recommend to use default which promises to extract reasonable
+        foreground objects.
+
+    percentile: int, range [0 100], default=80, optional
+        specifically used in threshold strategy which selects features
+        based on percentile value.
+
+    extractor: {'autoextractor', 'randomwalker'}, optional
+        switches between which technique is to be used. If 'autoextractor',
+        regions are extracted using a "Auto Extractor" method. Otherwise
+        if 'randomwalker', regions are extracted using "Random Walker"
+        segmentation algorithm. Both ideas are briefly defined above.
+
+    seed_ratio: float or string, default string {'auto'}. optional
+
+    smooth_fwhm: scalar, default smooth_fwhm=6. optional
+        smoothing parameter as a full width half maximum, in millimetres.
+        this scalar value is applied on all three directions.
+
+    Returns
+    -------
+    regions_: a 4D Nifti like image/object
+        a list of seperate regions with each region lying on a 3D volume
+        concatenated into a 4D Nifti object.
+
+    index_: a numpy array
+        an array of list of indices where each index value is assigned to
+        each seperate region of its corresponding family of brain maps.
+
+    References
+    ----------
+    * Abraham et al. "Region segmentation for sparse decompositions: better
+      brain parcellations from rest fMRI", Sparsity Techniques in Medical Imaging,
+      Sep 2014, Boston, United States. pp.8
+    """
+    def __init__(self, n_regions=None, min_size=20,
+                 threshold='auto', percentile=80,
+                 extractor='autoextractor', seed_ratio='auto',
+                 smooth_fwhm=6., verbose=0):
+        self.n_regions = n_regions
+        self.min_size = min_size
+        self.threshold = threshold
+        self.percentile = percentile
+        self.extractor = extractor
+        self.seed_ratio = seed_ratio
+        self.smooth_fwhm = smooth_fwhm
+        self.verbose = verbose
+
+    def transform(self, maps_img):
+        """ Extract the regions from the maps
+
+        Parameters
+        ----------
+        maps_img: a Nifti like image/object or the filename
+            the image consists of atlas maps or statistical maps
+
+        Returns
+        -------
+        regions_: 4D Nifti like image/object
+            a list of seperate regions concatenated in a 4D Nifti
+            like image.
+
+        index_: a numpy array
+            a list of indices in an array to get the identity of each
+            region to a family of brain atlas maps.
+        """
+        min_size = self.min_size
+        maps_img = check_niimg(maps_img)
```

Comment:
> I think that we can be more restrictive using check_niimg_4d.
> @GaelVaroquaux : people will want to extract regions from simple 3d images too
> but, in that case, thresholding is not very meaningful. Should we create
> another object or do a special case here?

Probably another object, but we need to discuss it. Basically, the
question is: how will we document / explain this?